tests/ui/mismatched_types/issue-118510.stderr

```diff
@@ -0,0 +1,52 @@
+error[E0412]: cannot find type `S` in this scope
+  --> $DIR/issue-118510.rs:2:19
+   |
+LL |     Ident(&'a mut S),
+   |                   ^ not found in this scope
+   |
+help: you might be missing a type parameter
+   |
+LL | pub enum Sexpr<'a, S> {
+   |                  +++
+
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/issue-118510.rs:6:23
+   |
+LL | fn map<'a, F: FnOnce(&Foo<'a>) -> T>(f: F) {}
+   |                       ^^^ not found in this scope
+
+error[E0412]: cannot find type `T` in this scope
+  --> $DIR/issue-118510.rs:6:35
+   |
+LL | fn map<'a, F: FnOnce(&Foo<'a>) -> T>(f: F) {}
+   |            -                      ^ help: a type parameter with a similar name exists: `F`
+   |            |
+   |            similarly named type parameter `F` defined here
+
+error[E0631]: type mismatch in function arguments
+  --> $DIR/issue-118510.rs:11:9
+   |
+LL |     Ident(&'a mut S),
+   |     ----- found signature defined here
+...
+LL |     map(Sexpr::Ident);
+   |     --- ^^^^^^^^^^^^ expected due to this
+   |     |
+   |     required by a bound introduced by this call
+   |
+   = note: expected function signature `for<'a> fn(&'a {type error}) -> _`
+              found function signature `fn(&mut {type error}) -> _`
+note: required by a bound in `map`
+  --> $DIR/issue-118510.rs:6:15
+   |
+LL | fn map<'a, F: FnOnce(&Foo<'a>) -> T>(f: F) {}
+   |               ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `map`
+help: consider wrapping the function in a closure
+   |
+LL |     map(|arg0: &{type error}| Sexpr::Ident(&mut *arg0));
```

Comment:
I'm wondering if it'd be better to not suggest a type annotation in the case of a type error--maybe something like `|arg0: /* type */|` instead in cases like these rather than `{type error}`?

Reply:
I don't think it matters much, you need to have a second entirely unrelated error to hit this case, which is quite unlikely in practice.
It happens in this test because the test is incomplete (https://github.com/rust-lang/rust/pull/118526#discussion_r1412785442).